opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential/aggregation.py

```diff
@@ -0,0 +1,555 @@
+            buckets = self._positive
+        else:
+            value = -value
+            buckets = self._negative
+
+        self._update(buckets, value, incr)
+
+    def _downscale(self, change: int) -> None:
+        """
+        Subtracts change from the current mapping scale
+        """
+
+        if change == 0:
+            return
+
+        if change < 0:
+            raise Exception(f"Impossible change of scale: {change}")
+
+        new_scale = self._mapping.scale - change
+
+        self._positive.downscale(change)
+        self._negative.downscale(change)
+
+        if new_scale <= 0:
+            mapping = ExponentMapping(new_scale)
+        else:
+            mapping = LogarithmExponentialHistogramMapping(new_scale)
+
+        self._mapping = mapping
+
+    # pylint: disable=no-self-use
+    def _change_scale(self, high: int, low: int, size: int) -> int:
+        """
+        Calculates how much downscaling is needed by shifting the high and low
+        values until they are separated by no more than size.
+        """
+
+        change = 0
+
+        # FIXME this slightly different from the Go implementation. It should
+        # be functionally equal but avoids an infinite loop in certain
+        # circumstances (high == 0, low == -1, size == 1).
+        while high - low > size:
+            high = high >> 1
+            low = low >> 1
+
+            change += 1
+
+        if high - low == size:
+            high = high >> 1
+            low = low >> 1
+
+            change += 1
+
+        return change
+
+    def _update(self, buckets: Buckets, value: float, incr: int) -> None:
+
+        index = self._mapping.map_to_index(value)
+
+        low, high, success = self._increment_index_by(buckets, index, incr)
+
+        if success:
+            return
+
+        self._downscale(self._change_scale(high, low, self._max_size))
+
+        index = self._mapping.map_to_index(value)
+
+        _, _, success = self._increment_index_by(buckets, index, incr)
+
+        if not success:
+            raise Exception("Downscale logic error")
+
+    def _increment_index_by(
+        self, buckets: Buckets, index: int, incr: int
+    ) -> tuple:
+        """
+        Determines if the index lies inside the current range
+        [indexStart, indexEnd] and, if not, returns the minimum size (up to
+        maxSize) will satisfy the new value.)]
+
+        Returns a tuple: low, high, success
+        """
+
+        if incr == 0:
+            # Skipping a bunch of work for 0 increment.  This
+            # happens when merging sparse data, for example.
+            # This also happens UpdateByIncr is used with a 0
+            # increment, means it can be safely skipped.
+
+            return 0, 0, True
+
+        if buckets.len() == 0:
+            # Go initializes its backing here if it hasn't been done before.
+            # I think we don't need to worry about that because the backing
+            # has been initialized already.
```

Comment:
```suggestion
            # Go initializes its backing here if it hasn't been done before.
            # I think we don't need to worry about that because the backing
            # has been initialized already.
```
Sounds good.  It leaves the backing `nil` until the first value arrives, but has no great reason for doing so.